cmd/influxd/restore.go

```diff
@@ -0,0 +1,257 @@
+type RestoreCommand struct {
+	// The logger passed to the ticker during execution.
+	Logger *log.Logger
+
+	// Standard input/output, overridden for testing.
+	Stderr io.Writer
+}
+
+// NewRestoreCommand returns a new instance of RestoreCommand with default settings.
+func NewRestoreCommand() *RestoreCommand {
+	return &RestoreCommand{
+		Stderr: os.Stderr,
+	}
+}
+
+// Run excutes the program.
+func (cmd *RestoreCommand) Run(args ...string) error {
+	// Set up logger.
+	cmd.Logger = log.New(cmd.Stderr, "", log.LstdFlags)
+	cmd.Logger.Printf("influxdb restore, version %s, commit %s", version, commit)
+
+	// Parse command line arguments.
+	config, path, err := cmd.parseFlags(args)
+	if err != nil {
+		return err
+	}
+
+	// Remove broker & data directories.
+	if err := os.RemoveAll(config.BrokerDir()); err != nil {
+		return fmt.Errorf("remove broker dir: %s", err)
+	} else if err := os.RemoveAll(config.DataDir()); err != nil {
+		return fmt.Errorf("remove data dir: %s", err)
+	}
+
+	// Open snapshot file.
+	f, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("open: %s", err)
+	}
+	defer f.Close()
+
+	// Create reader and extract manifest.
+	sr := influxdb.NewSnapshotReader(f)
+	ss, err := sr.Snapshot()
+	if err != nil {
+		return fmt.Errorf("snapshot: %s", err)
+	}
+
+	// Unpack snapshot files into data directory.
+	if err := cmd.unpack(config.DataDir(), sr); err != nil {
+		return fmt.Errorf("unpack: %s", err)
+	}
+
+	// Generate broker & raft directories from manifest.
+	if err := cmd.materialize(config.BrokerDir(), ss, config.BrokerURL()); err != nil {
+		return fmt.Errorf("materialize: %s", err)
+	}
+
+	// Notify user of completion.
+	cmd.Logger.Println("restore complete")
+
+	return nil
+}
+
+// parseFlags parses and validates the command line arguments.
+func (cmd *RestoreCommand) parseFlags(args []string) (*Config, string, error) {
+	fs := flag.NewFlagSet("", flag.ContinueOnError)
+	configPath := fs.String("config", "", "")
+	fs.SetOutput(cmd.Stderr)
+	fs.Usage = cmd.printUsage
+	if err := fs.Parse(args); err != nil {
+		return nil, "", err
+	}
+
+	// Parse configuration file from disk.
+	config, err := parseConfig(*configPath, "")
+	if err != nil {
+		log.Fatal(err)
+	} else if *configPath == "" {
+		log.Println("No config provided, using default settings")
+	}
+
+	// Require output path.
+	path := fs.Arg(0)
+	if path == "" {
+		return nil, "", fmt.Errorf("snapshot path required")
+	}
+
+	return config, path, nil
+}
+
+// unpack expands the files in the snapshot archive into a directory.
+func (cmd *RestoreCommand) unpack(path string, sr *influxdb.SnapshotReader) error {
+	// Create root directory.
+	if err := os.MkdirAll(path, 0777); err != nil {
+		return fmt.Errorf("mkdir: err=%s", err)
+	}
+
+	// Loop over files and extract.
+	for {
+		// Read entry header.
+		sf, err := sr.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return fmt.Errorf("next: entry=%s, err=%s", sf.Name, err)
+		}
+
```

Comment:
Should it log output here just to let the user know that the file is being done? That way they get incremental output as the restore progresses